src/java/htsjdk/samtools/SRAIterator.java

```diff
@@ -0,0 +1,195 @@
+*  to the public for use. The National Library of Medicine and the U.S.
+*  Government have not placed any restriction on its use or reproduction.
+*
+*  Although all reasonable efforts have been taken to ensure the accuracy
+*  and reliability of the software and data, the NLM and the U.S.
+*  Government do not and cannot warrant the performance or results that
+*  may be obtained by using this software or data. The NLM and the U.S.
+*  Government disclaim all warranties, express or implied, including
+*  warranties of performance, merchantability or fitness for any particular
+*  purpose.
+*
+*  Please cite the author in any work or product based on this material.
+*
+* ===========================================================================
+*
+*/
+
+/**
+ * Created by andrii.nikitiuk on 8/11/15.
+ */
+
+package htsjdk.samtools;
+
+import htsjdk.samtools.SAMFileHeader.SortOrder;
+
+import htsjdk.samtools.sra.SRAAlignmentIterator;
+import htsjdk.samtools.sra.SRAUnalignmentIterator;
+import ngs.ErrorMsg;
+import ngs.ReadCollection;
+import ngs.Reference;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public class SRAIterator implements SAMRecordIterator {
+    private ValidationStringency validationStringency;
+
+    private ReadCollection run;
+    private SAMFileHeader header;
+    private final List<Reference> cachedReferences;
+    private RecordRangleInfo recordRangleInfo;
+    private Iterator<Chunk> chunksIterator;
+    private Chunk currentChunk;
+
+    private SRAAlignmentIterator alignmentIterator;
+    private SRAUnalignmentIterator unalignmentIterator;
+
+    public static class RecordRangleInfo {
+        private List<Long> referenceOffsets;
+        private List<Long> referenceLengths;
+        private long totalReferencesLength;
+        private long numberOfReads; // is used for unaligned read space
+        private long totalRecordRangeLength;
+
+        public RecordRangleInfo(List<Long> referenceLengths, long numberOfReads) {
+            this.numberOfReads = numberOfReads;
+            this.referenceLengths = referenceLengths;
+
+            referenceOffsets = new ArrayList<Long>();
+
+            totalReferencesLength = 0;
+            for (Long refLen : referenceLengths) {
+                referenceOffsets.add(totalReferencesLength);
+                totalReferencesLength += refLen;
+            }
+
+            totalRecordRangeLength = totalReferencesLength + this.numberOfReads;
+        }
+
+        public long getNumberOfReads() {
+            return numberOfReads;
+        }
+
+        public long getTotalReferencesLength() {
+            return totalReferencesLength;
+        }
+
+        public long getTotalRecordRangeLength() {
+            return totalRecordRangeLength;
+        }
+
+        public final List<Long> getReferenceOffsets() {
+            return referenceOffsets;
```

Comment:
Should the caller be able to modify the returned `List`? If not, wrap in an `UnmodifiableList` (here and below)